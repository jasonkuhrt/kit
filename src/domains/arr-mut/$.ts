export * as ArrMut from './$$.ts'
