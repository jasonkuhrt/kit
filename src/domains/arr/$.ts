/**
 * Array utilities for working with readonly and mutable arrays.
 *
 * Provides functional utilities for array operations including mapping, filtering,
 * type guards, and conversions. Emphasizes immutable operations and type safety.
 *
 * @category Core Data Structures
 */
export * as Arr from './$$.js'
