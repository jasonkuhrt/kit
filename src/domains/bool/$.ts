export * as Bool from './$$.js'
