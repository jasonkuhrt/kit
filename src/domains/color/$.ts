// @ts-expect-error Duplicate identifier
export * as Color from './$$.js'

/**
 * Color utilities for terminal output.
 *
 * Provides comprehensive color support for CLI applications including:
 * - **140 CSS Level 4 named colors** (red, blue, etc.)
 * - **Multiple input formats**: hex, RGB, HSL, CSS functions, and objects
 * - **RGB (24-bit true color)** as canonical format
 * - **Effect Schema validation** with branded types
 * - **Type-level parser** for compile-time hints
 * - **Color conversions** between formats
 *
 * ## Supported Formats
 *
 * - Hex: `#FF5733` or `FF5733`
 * - RGB space-separated: `rgb 255 87 51`
 * - RGB CSS function: `rgb(255, 87, 51)`
 * - HSL space-separated: `hsl 120 100 50`
 * - HSL CSS function: `hsl(120, 100, 50)`
 * - Named colors: `red`, `blue`, `green`, etc. (140 colors)
 * - Object: `{ r: 255, g: 87, b: 51 }`
 *
 * ## Color Properties
 *
 * Colors can be applied to three targets:
 * - **foreground**: Text color
 * - **background**: Background color
 * - **underlineColor**: Underline color (when used with underline modifier)
 *
 * ## Usage
 *
 * @example
 * ```typescript
 * import { Color } from '@wollybeard/kit/color'
 *
 * // Parse various formats
 * const red = Color.fromString('red')
 * const blue = Color.fromString('#0000FF')
 * const green = Color.fromString('rgb 0 255 0')
 * const yellow = Color.fromString('hsl(60, 100, 50)')
 *
 * // Convert to hex
 * red.toHex()  // '#FF0000'
 *
 * // Use in schemas
 * const ConfigSchema = S.Struct({
 *   primaryColor: Color.String,
 *   secondaryColor: ColorInput
 * })
 * ```
 *
 * @category Color
 */
export namespace Color {}
