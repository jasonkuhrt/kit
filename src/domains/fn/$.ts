/**
 * Function utilities for functional programming patterns.
 *
 * Provides utilities for function composition, currying, partial application,
 * piping, and common functional patterns like identity and constant functions.
 * Includes function analysis tools and endomorphism utilities.
 *
 * @category Core Data Structures
 */
export * as Fn from './$$.js'
