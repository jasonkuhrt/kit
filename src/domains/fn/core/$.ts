export * as CoreFn from './$$.js'
