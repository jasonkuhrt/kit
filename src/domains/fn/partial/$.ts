export * as Partial from './$$.js'
