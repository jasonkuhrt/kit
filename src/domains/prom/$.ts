/**
 * Promise utilities for asynchronous operations.
 *
 * Provides utilities for working with Promises including deferred promise
 * creation, promise combinators, and async control flow patterns.
 *
 * @category Error Handling & Values
 */
export * as Prom from './$$.js'
