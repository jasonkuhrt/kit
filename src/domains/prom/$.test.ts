import { property } from '#test/test'
import fc from 'fast-check'
import { expect, test } from 'vitest'
import { Prom } from './$.js'

test('isShape detects promises and thenables', () => {
  expect(Prom.isShape(Promise.resolve(42))).toBe(true)
  // eslint-disable-next-line unicorn/no-thenable
  expect(Prom.isShape({ then: () => {}, catch: () => {}, finally: () => {} })).toBe(true)
  // eslint-disable-next-line unicorn/no-thenable
  expect(Prom.isShape({ then: () => {} })).toBe(false)
  // eslint-disable-next-line unicorn/no-thenable
  expect(Prom.isShape({ then: 'not a function', catch: () => {}, finally: () => {} })).toBe(false)
})

property(
  'isShape rejects non-promise values',
  fc.oneof(
    fc.integer(),
    fc.string(),
    fc.boolean(),
    fc.constant(null),
    fc.constant(undefined),
    fc.object(),
  ),
  (value) => {
    expect(Prom.isShape(value)).toBe(false)
  },
)
