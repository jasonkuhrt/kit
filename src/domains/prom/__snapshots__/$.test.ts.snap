// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`createDeferred > state tracking > after reject 1`] = `
"
╔══════════════════════════════════════════════════╗ RUNNER
d.reject(new Error("test"));
d.promise.catch(() => {
});
╠══════════════════════════════════════════════════╣ OUTPUTS RETURN
{
  "promise": {},
  "isResolved": false,
  "isRejected": true,
  "isSettled": true
}
╚══════════════════════════════════════════════════╝"
`;

exports[`createDeferred > state tracking > after resolve 1`] = `
"
╔══════════════════════════════════════════════════╗ RUNNER
d.resolve(1);
╠══════════════════════════════════════════════════╣ OUTPUTS RETURN
{
  "promise": {},
  "isResolved": true,
  "isRejected": false,
  "isSettled": true
}
╚══════════════════════════════════════════════════╝"
`;

exports[`createDeferred > state tracking > initial 1`] = `
"
╔══════════════════════════════════════════════════╗ RUNNER

╠══════════════════════════════════════════════════╣ OUTPUTS RETURN
{
  "promise": {},
  "isResolved": false,
  "isRejected": false,
  "isSettled": false
}
╚══════════════════════════════════════════════════╝"
`;

exports[`createDeferred > strict > reject resolve calls throw 1`] = `
"
╔══════════════════════════════════════════════════╗ RUNNER
d.reject(new Error("test"));
d.promise.catch(() => {
});
d.resolve(1);
╠══════════════════════════════════════════════════╣ OUTPUTS THROW
Error: Deferred promise already settled
╚══════════════════════════════════════════════════╝"
`;

exports[`createDeferred > strict > resolve reject calls throw 1`] = `
"
╔══════════════════════════════════════════════════╗ RUNNER
d.reject(new Error("test"));
d.promise.catch(() => {
});
d.resolve(1);
╠══════════════════════════════════════════════════╣ OUTPUTS THROW
Error: Deferred promise already settled
╚══════════════════════════════════════════════════╝"
`;

exports[`createDeferred > strict > resolve resolve calls throw 1`] = `
"
╔══════════════════════════════════════════════════╗ RUNNER
d.resolve(1);
d.resolve(1);
╠══════════════════════════════════════════════════╣ OUTPUTS THROW
Error: Deferred promise already settled
╚══════════════════════════════════════════════════╝"
`;
